### DISCUSSION TITLE
Memory issues when fine-tuning Qwen2-VL with LoRA on 16GB M1 Max - any optimization tips?

### DISCUSSION THREAD

**Discussion by BittuChan** [Blaizzy/mlx-vlm#932] [Q&A] [answered]
Hello everyone, 

I've been trying to fine-tune the Qwen2-VL-7B model on my M1 Max (16GB unified memory) using the LoRA setup from the docs, but keep hitting OOM errors during training. I'm working with a custom dataset of around 800 image-text pairs for a document classification task. Here's my current command:

```
mlx_vlm.lora --model mlx-community/Qwen2-VL-7B-Instruct-4bit \
  --data ./data \
  --batch-size 2 \
  --lora-layers 16 \
  --iters 1000
```

The model loads fine and inference works perfectly, but training crashes about 50 iterations in with a memory error. I've already tried:
1. Reducing batch size to 1, this helps a bit but still crashes eventually
2. Using the 4-bit quantized model which I already am
3. Closing all other apps - Chrome was definitely hogging memory

A few specific questions:
- Is there a practical limit for VLM fine-tuning on 16GB? Should I just accept that 7B models are too big for this setup?
- Also I noticed there's a --grad-checkpoint flag in mlx-lm but didn't see it in mlx-vlm docs.

For context, I'm targeting a research paper submission so I need this to actually work, but I don't have access to beefier hardware right now. Any advice from folks who've successfully fine-tuned VLMs on similar setups would be super helpful.
Thanks!

**Comment by Shiven0504** [comment, marked as answer]
hii @BittuChan, I've previously used fine-tuning VLMs on a 16GB M2 Pro. Here's what worked for me: 

Quick answer: Yes, 7B VLMs are tight on 16GB, but totally doable with the right settings.

Memory Optimization Strategies
1. Reduce LoRA Layers
This is your biggest lever. Dropping from 16 to 8 layers cuts your trainable parameters roughly in half:

```
mlx_vlm.lora --model mlx-community/Qwen2-VL-7B-Instruct-4bit \
  --data ./data \
  --batch-size 1 \
  --lora-layers 8 \  # Try 8 first, then 4 if still crashing
  --iters 1000
```

For document classification (which sounds like a focused task), 8 layers should be plenty. I've seen good results with as few as 4 layers for similar use cases. The quality difference isn't massive unless you're doing something really complex.

2. Gradient Checkpointing
MLX-VLM does support gradient checkpointing (it's in the codebase but maybe not well documented). Try adding:
`--grad-checkpoint`
This trades compute for memory by recomputing activations during backward pass instead of storing them. Slows training by maybe 10-15% but can save 2-3GB.

3. Sequence Length Matters
If your images are getting resized to high resolutions, that eats memory fast. Try limiting:
`--max-seq-length 2048  # or even 1024 for initial experiments`
Also check your image preprocessing. If you're using massive input images, downscale them during the data prep step.

**Reply by Shiven0504** [reply]
Also I would suggest you monitor memory usage with `sudo powermetrics --samplers smc -i 1000` during training to see exactly when you're hitting the limit. If you're using the `[torch]` extra dependencies for preprocessing, make sure you're not keeping torch tensors in memory unnecessarily. Some models have a `max-pixels` parameter that controls image resolution so perhaps reducing this can help (e.g., --max-pixels 224 224)

Hope this helps! Let me know which combo works for you.

**Reply by BittuChan** [reply]
Hey @Shiven0504 , thanks for your suggestion! I think I'll go with:
- `--lora-layers 8`
- `--batch-size 1`
- `--grad-checkpoint`
- `--max-seq-length 1024`

Its running smoothly now, the big realization for me was that more LoRA layers ≠ better results for my specific task. I ran a quick eval after 200 iterations and the loss curve looks solid even with just 8 layers.

**Comment by BittuChan** [comment]
Also turns out my dataset was preprocessing images at full resolution (some were 4K+) which was definitely contributing to the OOM issues. Capped them at 768x768 and that helped a ton.

Marking this as resolved. Really appreciate the detailed breakdown, this saved me days of trial and error!
